File: backend/src/Buscador.Domain/Animais/Animal.cs
using Buscador.Domain.Comum;

namespace Buscador.Domain.Animais;

public sealed class Animal : RaizAgregada<AnimalId>
{
    public string NomeComum { get; private set; } = string.Empty;
    public string NomeCientifico { get; private set; } = string.Empty;
    public string Descricao { get; private set; } = string.Empty;
    public string Caracteristicas { get; private set; } = string.Empty;
    public Dieta Dieta { get; private set; }
    public Habitat Habitat { get; private set; }
    public string DistribuicaoGeografica { get; private set; } = string.Empty;
    public StatusConservacao StatusConservacao { get; private set; }
    public string[] Tags { get; private set; } = [];
    public string Curiosidades { get; private set; } = string.Empty;

    private Animal() { }

    private Animal(AnimalId id) : base(id) { }

    public static Animal Criar(
        string nomeComum,
        string nomeCientifico,
        string descricao,
        string caracteristicas,
        Dieta dieta,
        Habitat habitat,
        string distribuicaoGeografica,
        StatusConservacao statusConservacao,
        string[] tags,
        string curiosidades)
    {
        if (string.IsNullOrWhiteSpace(nomeComum))
            throw new ArgumentException("Nome comum não pode ser vazio.", nameof(nomeComum));

        if (string.IsNullOrWhiteSpace(nomeCientifico))
            throw new ArgumentException("Nome científico não pode ser vazio.", nameof(nomeCientifico));

        return new Animal(AnimalId.Novo())
        {
            NomeComum = nomeComum,
            NomeCientifico = nomeCientifico,
            Descricao = descricao,
            Caracteristicas = caracteristicas,
            Dieta = dieta,
            Habitat = habitat,
            DistribuicaoGeografica = distribuicaoGeografica,
            StatusConservacao = statusConservacao,
            Tags = tags,
            Curiosidades = curiosidades
        };
    }
}

File: frontend/src/types/animal.ts
export type ModoBusca = "Textual" | "Semantica" | "Hibrida";

export interface Animal {
  id: string;
  nomeComum: string;
  nomeCientifico: string;
  descricao: string;
  caracteristicas: string;
  dieta: number;
  habitat: number;
  distribuicaoGeografica: string;
  statusConservacao: number;
  tags: string[];
  curiosidades: string;
}

export interface ResultadoBusca {
  animal: Animal;
  pontuacao: number;
}

// Mapas int -> rotulo PT, derivados dos enums do Domain (.NET). A ordem dos indices
// segue EXATAMENTE a ordem de declaracao dos enums em backend/src/Buscador.Domain/Animais/.
export const rotulosDieta: Record<number, string> = {
  0: "Carnivoro",
  1: "Herbivoro",
  2: "Onivoro",
};

export const rotulosHabitat: Record<number, string> = {
  0: "Floresta",
  1: "Oceano",
  2: "Deserto",
  3: "Savana",
  4: "Montanha",
  5: "Agua doce",
  6: "Polar",
};

export const rotulosStatusConservacao: Record<number, string> = {
  0: "Pouco preocupante",
  1: "Quase ameacado",
  2: "Vulneravel",
  3: "Em perigo",
  4: "Criticamente em perigo",
  5: "Extinto na natureza",
  6: "Extinto",
};
